### PR TITLE
HOTT-2119: Improve parts of speech handling when doing synonym search

### PR DESCRIPTION
**Pull request by willfish**
### Jira link

https://transformuk.atlassian.net/browse/HOTT-2119

### What?

We need to make sure we're correctly using the synonym engine in the opensearch cluster.

Currently, we'll try and tokenise and correct the spelling of the input search query and this adjusted query doesn't execute the synonym engine. 

You can see this when looking at the result of the parsed synonym `ash trees` below:

```shell
# Do a search with non-parts-of-speech input using the english analyzer (synonym match)


curl --silent 'https://LAHPa6z9:<redacted>@tariff-search-development.trade-tariff.service.gov.uk/tariff-goods_nomenclatures-uk/_analyze' -d '{"analyzer":"english","text":"ash trees"}' -H 'Content-Type: application/json'
{"tokens":[{"token":"tree","start_offset":0,"end_offset":9,"type":"SYNONYM","position":0}]}


# Do a search with the parts-of-speech input using the english analyzer (no synonym match)

curl --silent 'https://LAHPa6z9:<redacted>@tariff-search-development.trade-tariff.service.gov.uk/tariff-goods_nomenclatures-uk/_analyze' -d '{"analyzer":"english","text":"ash tree"}' -H 'Content-Type: application/json'
{"tokens":[{"token":"ash","start_offset":0,"end_offset":3,"type":"<ALPHANUM>","position":0},{"token":"tree","start_offset":4,"end_offset":8,"type":"<ALPHANUM>","position":1}]}
```

We want the resulting tokens generated for ash trees and not for ash tree.

The approach I've picked to solve this is to 

1. Persist the aggregated synonyms in the live docker container
2. Build a Set to check in constant time whether there are any aggregated synonyms that are a phrase match on our input search query
3. When we do match we bypass generating parts of speech and instead use essentially a null object pattern for the synonym search query result and therefore force the use of the synonym functionality in opensearch.

I have added/removed/altered:

- [x] Adds aggregate_synonyms bin script
- [x] Generate aggregated synonyms on build
- [x] Support synonym parser result
- [x] Simplify goods nomenclature item id method
- [x] Adds synonym matcher service

### Why?

I am doing this because:

- This is required to avoid splitting the parts of speech/doing any stemming of the input query and instead relying on the synonym engine in the opensearch cluster

### CODE REVIEW

**Review by jebw — CHANGES_REQUESTED** [review]
Couple of comments
* theres a misspelling but the implementation is a little quirky anyway
* the presence of the synonyms file seems to be needed in the docker image so should be created as part of the build process?

File: app/lib/trade_tariff_backend.rb [hunk 1]
```diff
@@ -270,5 +270,9 @@ def beta_search_debug?
     def use_uk_xi_indexes?
       ENV['USE_UK_XI_INDEXES'] == 'true'
     end
+
+    def aggregated_synonyms_file
```
[the remark this answers is not in the document]
Comment (willfish) [reply app/lib/trade_tariff_backend.rb:274]: Anything that comes from the environment I put in the trade_tariff_backend module so this is fine

File: app/models/beta/search/search_query_parser_result.rb [hunk 2]
```diff
@@ -2,31 +2,52 @@ module Beta
...
       attr_accessor :adjectives,
                     :nouns,
                     :verbs,
                     :noun_chunks,
                     :original_search_query,
-                    :corrected_search_query
+                    :corrected_search_query,
+                    :synonym_result
+
+      alias_method :syonym_result?, :synonym_result
```
[the remark this answers is not in the document]
Comment (willfish) [reply app/models/beta/search/search_query_parser_result.rb:17]: Oh interesting spot. I'll look into it

File: .circleci/config.yml [hunk 3]
```diff
@@ -342,6 +342,10 @@ jobs:
           name: "Set docker tag"
           command: |
             echo "export DOCKER_TAG=<<# parameters.dev-build >>dev-<</ parameters.dev-build >>${CIRCLE_SHA1}" >> $BASH_ENV
+
+      - run:
+          name: "Aggregate synonyms"
+          command: bin/aggregate_synonyms
```
[the remark this answers is not in the document]
Comment (willfish) [reply .circleci/config.yml:348]: So we _could_ add it to the Dockerfile. The synonym package management happens outside of the build step, though, so we'll still have duplicating script execution

**Review by jebw — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**